In [43]:
import pandas as pd

In [44]:
df = pd.read_csv('cars24-car-price-cleaned.csv')

* We want to convert the categorical data like make, model into categorical data

In [45]:
df['make'].unique()

array(['MARUTI', 'HYUNDAI', 'FORD', 'MAHINDRA', 'TATA', 'RENAULT',
       'NISSAN', 'MINI', 'MERCEDES-BENZ', 'TOYOTA', 'FIAT', 'VOLKSWAGEN',
       'HONDA', 'CHEVROLET', 'AMBASSADOR', 'DATSUN', 'KIA', 'BMW',
       'MITSUBISHI', 'AUDI', 'SKODA', 'LAND', 'JAGUAR', 'DAEWOO',
       'BENTLEY', 'MG', 'ISUZU', 'PORSCHE', 'VOLVO', 'LEXUS', 'JEEP',
       'PREMIER', 'MASERATI', 'FORCE', 'LAMBORGHINI', 'FERRARI',
       'OPELCORSA', 'MERCEDES-AMG', 'DC', 'ROLLS-ROYCE', 'OPEL'],
      dtype=object)

In [46]:
df.groupby('make')['selling_price'].mean().to_dict()

{'AMBASSADOR': 1.4525000000000001,
 'AUDI': 17.177584876543193,
 'BENTLEY': 20.90249999999996,
 'BMW': 17.34901656314698,
 'CHEVROLET': 2.7236206896551725,
 'DAEWOO': 0.78,
 'DATSUN': 3.2051764705882353,
 'DC': 11.925,
 'FERRARI': 20.90249999999996,
 'FIAT': 3.1527951327433628,
 'FORCE': 5.5200000000000005,
 'FORD': 5.858258333333332,
 'HONDA': 5.9799016301292856,
 'HYUNDAI': 5.4588194834362715,
 'ISUZU': 14.03770833333333,
 'JAGUAR': 20.187499999999968,
 'JEEP': 16.261890243902435,
 'KIA': 15.731515151515147,
 'LAMBORGHINI': 20.90249999999996,
 'LAND': 20.560147058823492,
 'LEXUS': 20.90249999999996,
 'MAHINDRA': 7.315421238244514,
 'MARUTI': 4.684721460176991,
 'MASERATI': 20.90249999999996,
 'MERCEDES-AMG': 20.90249999999996,
 'MERCEDES-BENZ': 17.07147938144328,
 'MG': 17.529473684210526,
 'MINI': 18.52945652173911,
 'MITSUBISHI': 8.63102564102564,
 'NISSAN': 4.611877162629757,
 'OPEL': 1.5,
 'OPELCORSA': 1.5166666666666666,
 'PORSCHE': 20.866399999999963,
 'PREMIER': 2.55,
 'RENAUL

* As we see LAMBORGHINI, LEXUS have the same means (20.902500, 20.902500)
* We can iterate on values that have same means and change them by precision, so that the make has it's own representation

In [47]:
#  grouping by make and taking mean of it's Selling price, to replace each value with mean of the group
df['make'] = df.groupby('make')['selling_price'].transform('mean')
df['model'] = df.groupby('model')['selling_price'].transform('mean')

* The 'make' and 'model' has been coverted into means (numerical) of it's make group selling price

In [48]:
df.describe()

selling_price          year     km_driven       mileage        engine  \
count   19820.000000  19820.000000  1.982000e+04  19820.000000  19820.000000   
mean        6.585509   2014.561453  5.815856e+04     19.503402   1475.702381   
std         4.847364      3.196636  5.171563e+04      4.297784    518.571223   
min         0.300000   1992.000000  1.000000e+02      4.000000      0.000000   
25%         3.410000   2013.000000  3.100000e+04     16.950000   1197.000000   
50%         5.200000   2015.000000  5.200000e+04     19.300000   1248.000000   
75%         7.850000   2017.000000  7.400000e+04     22.320000   1582.000000   
max        20.902500   2021.000000  3.800000e+06    120.000000   6752.000000   

          max_power           age          make         model    Individual  \
count  19820.000000  19820.000000  19820.000000  19820.000000  19820.000000   
mean      98.122907      8.438547      6.585509      6.585509      0.390666   
std       44.761727      3.196636      3.633358      4.692537      0.487912   
min        5.000000      2.000000      0.780000      0.325000      0.000000   
25%       73.900000      6.000000      4.684721      3.477742      0.000000   
50%       86.800000      8.000000      5.458819      5.210816      0.000000   
75%      112.000000     10.000000      5.979902      7.661538      1.000000   
max      626.000000     31.000000     20.902500     20.902500      1.000000   

       Trustmark Dealer        Diesel      Electric           LPG  \
count      19820.000000  19820.000000  19820.000000  19820.000000   
mean           0.009586      0.492583      0.000404      0.003229   
std            0.097442      0.499958      0.020087      0.056734   
min            0.000000      0.000000      0.000000      0.000000   
25%            0.000000      0.000000      0.000000      0.000000   
50%            0.000000      0.000000      0.000000      0.000000   
75%            0.000000      1.000000      0.000000      0.000000   
max            1.000000      1.000000      1.000000      1.000000   

             Petrol        Manual             5            >5  
count  19820.000000  19820.000000  19820.000000  19820.000000  
mean       0.487841      0.802674      0.835015      0.152825  
std        0.499865      0.397990      0.371176      0.359828  
min        0.000000      0.000000      0.000000      0.000000  
25%        0.000000      1.000000      1.000000      0.000000  
50%        0.000000      1.000000      1.000000      0.000000  
75%        1.000000      1.000000      1.000000      0.000000  
max        1.000000      1.000000      1.000000      1.000000

* 3.800000e+06 The km_driven max is too high

In [49]:
df

selling_price    year  km_driven  mileage  engine  max_power   age  \
0               1.20  2012.0     120000    19.70   796.0      46.30  11.0   
1               5.50  2016.0      20000    18.90  1197.0      82.00   7.0   
2               2.15  2010.0      60000    17.00  1197.0      80.00  13.0   
3               2.26  2012.0      37000    20.92   998.0      67.10  11.0   
4               5.70  2015.0      30000    22.77  1498.0      98.59   8.0   
...              ...     ...        ...      ...     ...        ...   ...   
19815           6.50  2017.0      69480    23.59  1364.0      67.05   6.0   
19816           9.25  2019.0      18000    17.50  1373.0      91.10   4.0   
19817           4.25  2015.0      67000    21.14  1498.0     103.52   8.0   
19818          12.25  2016.0    3800000    16.00  2179.0     140.00   7.0   
19819          12.00  2019.0      13000    18.00  1497.0     117.60   4.0   

            make      model  Individual  Trustmark Dealer  Diesel  Electric  \
0       4.684721   1.180000           1                 0       0         0   
1       5.458819   4.818750           1                 0       0         0   
2       5.458819   3.394000           1                 0       0         0   
3       4.684721   2.242676           1                 0       0         0   
4       5.858258   6.777576           0                 0       1         0   
...          ...        ...         ...               ...     ...       ...   
19815  10.532763   7.075000           0                 0       1         0   
19816   4.684721   7.128571           0                 0       0         0   
19817   7.182097   4.454000           0                 0       1         0   
19818   7.315421   8.096522           0                 0       1         0   
19819   5.979902  11.014286           0                 0       0         0   

       LPG  Petrol  Manual  5  >5  
0        0       1       1  1   0  
1        0       1       1  1   0  
2        0       1       1  1   0  
3        0       1       1  1   0  
4        0       0       1  1   0  
...    ...     ...     ... ..  ..  
19815    0       0       1  1   0  
19816    0       1       1  0   1  
19817    0       0       1  1   0  
19818    0       0       1  0   1  
19819    0       1       0  1   0  

[19820 rows x 18 columns]

In [37]:
# Max kms_driven
df.loc[df['km_driven'] == 3.800000e+06]
# 38 lakh kms

selling_price    year  km_driven  mileage  engine  max_power  age  \
19818          12.25  2016.0    3800000     16.0  2179.0      140.0  7.0   

           make          model  Individual  Trustmark Dealer  Diesel  \
19818  MAHINDRA  XUV500 W6 2WD           0                 0       1   

       Electric  LPG  Petrol  Manual  5  >5  
19818         0    0       0       1  0   1

## Difference betwenn fit and transform
https://www.analyticsvidhya.com/blog/2021/04/difference-between-fit-transform-fit_transform-methods-in-scikit-learn-with-python-code/#

In [51]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
#data after transformation
data = pd.DataFrame(scaler.fit_transform(df),columns=df.columns)

selling_price      year  km_driven   mileage    engine  max_power  \
0          -1.111046 -0.801317   1.195828  0.045745 -1.310754  -1.157780   
1          -0.223944  0.450030  -0.737872 -0.140402 -0.537456  -0.360203   
2          -0.915058 -1.426990   0.035608 -0.582501 -0.537456  -0.404885   
3          -0.892365 -0.801317  -0.409143  0.329620 -0.921213  -0.693085   
4          -0.182683  0.137194  -0.544502  0.760085  0.042999   0.010435   
...              ...       ...        ...       ...       ...        ...   
19815      -0.017641  0.762867   0.218923  0.950886 -0.215410  -0.694202   
19816       0.549692  1.388540  -0.776546 -0.466159 -0.198054  -0.156899   
19817      -0.481822  0.137194   0.170967  0.380810  0.042999   0.120577   
19818       1.168601  0.450030  72.355997 -0.815185  1.356256   0.935579   
19819       1.117025  1.388540  -0.873231 -0.349818  0.041071   0.435139   

            age      make     model  Individual  Trustmark Dealer    Diesel  \
0      0.801317 -0.523162 -1.151966    1.248892         -0.098382 -0.985275   
1     -0.450030 -0.310104 -0.376513    1.248892         -0.098382 -0.985275   
2      1.426990 -0.310104 -0.680141    1.248892         -0.098382 -0.985275   
3      0.801317 -0.523162 -0.925500    1.248892         -0.098382 -0.985275   
4     -0.137194 -0.200164  0.040931   -0.800710         -0.098382  1.014945   
...         ...       ...       ...         ...               ...       ...   
19815 -0.762867  1.086420  0.104315   -0.800710         -0.098382  1.014945   
19816 -1.388540 -0.523162  0.115732   -0.800710         -0.098382 -0.985275   
19817 -0.137194  0.164202 -0.454245   -0.800710         -0.098382  1.014945   
19818 -0.450030  0.200897  0.322012   -0.800710         -0.098382  1.014945   
19819 -1.388540 -0.166684  0.943815   -0.800710         -0.098382 -0.985275   

       Electric       LPG    Petrol    Manual         5        >5  
0     -0.020095 -0.056917  1.024622  0.495818  0.444503 -0.424728  
1     -0.020095 -0.056917  1.024622  0.495818  0.444503 -0.424728  
2     -0.020095 -0.056917  1.024622  0.495818  0.444503 -0.424728  
3     -0.020095 -0.056917  1.024622  0.495818  0.444503 -0.424728  
4     -0.020095 -0.056917 -0.975970  0.495818  0.444503 -0.424728  
...         ...       ...       ...       ...       ...       ...  
19815 -0.020095 -0.056917 -0.975970  0.495818  0.444503 -0.424728  
19816 -0.020095 -0.056917  1.024622  0.495818 -2.249703  2.354446  
19817 -0.020095 -0.056917 -0.975970  0.495818  0.444503 -0.424728  
19818 -0.020095 -0.056917 -0.975970  0.495818 -2.249703  2.354446  
19819 -0.020095 -0.056917  1.024622 -2.016868  0.444503 -0.424728  

[19820 rows x 18 columns]

* Converts everything into 2d numpy array

In [52]:
y = df['selling_price']
x = df.drop(columns=['selling_price'])

In [54]:
from sklearn.model_selection import train_test_split

X_train, X_test,Y_train,Y_test= train_test_split(x,y, test_size = 0.33, random_state=42)

In [55]:
X_train.shape

(13279, 17)

In [56]:
X_test.shape

(6541, 17)

In [57]:
Y_train.shape

(13279,)

In [58]:
Y_test.shape

(6541,)